### PR TITLE
Add ChiSquared pdf

### PR DESCRIPTION
**Pull request by anjabeck**
## Proposed Changes
- add chi2 distribution from [tfp](https://www.tensorflow.org/probability/api_docs/python/tfp/distributions/Chi2)

## Tests added
- check chi2 mean and std deviation

## Checklist

- [x] change approved
- [x] implementation finished
- [x] correct namespace imported
- [x] tests added
- [x] CHANGELOG updated
- [x] (if new public method/class) docs in rst file updated?

related to #512

### CODE REVIEW

File: tests/test_pdf_chisquared.py
Comment (jonas-eschle) [inline tests/test_pdf_chisquared.py:11]: create the obs and test values inside the test function (and use znp.name instead of np.name wherever possible)

File: tests/test_pdf_chisquared.py
Comment (jonas-eschle) [inline tests/test_pdf_chisquared.py:7]: Use `zfit.pdf` to access the pdf. And remove the imports, inside the test, you can directly use `zfit.pdf.name` or `zfit.Parameter`

File: zfit/models/dist_tfp.py
Comment (jonas-eschle) [inline zfit/models/dist_tfp.py:514]: An empty line is needed between the directive and the content (yes, RST is quite sensitive to many things: https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-math)

File: zfit/models/dist_tfp.py
Comment (jonas-eschle) [inline zfit/models/dist_tfp.py:552]: Can you add here a Repr (as for the others)? This makes the PDF serializable

File: tests/test_pdf_chisquared.py
Comment (jonas-eschle) [inline tests/test_pdf_chisquared.py:21]:
```suggestion
```
Not needed, has no effect

File: tests/test_pdf_chisquared.py
Comment (jonas-eschle) [inline tests/test_pdf_chisquared.py:23]:
```suggestion
    samples = chi2.sample(10000)
```
`.numpy()` is not needed in general

**Review by jonas-eschle — CHANGES_REQUESTED** [review]
Thanks for the addition! A few minor things, but otherwise looks good to me.

Two more things:
- add the PDF to the docs, under docs/user_api/pdf/basic.rst
- add yourself to the contributors in AUTHORS.rst

**Comment by ikrommyd** [comment]
Please also test in `tests/serialize/test_basics.py`

File: tests/test_pdf_chisquared.py
Comment (anjabeck) [reply tests/test_pdf_chisquared.py:11]: Ok sure.
For my curiosity: they are global because I started from a copy of [test_pdf_poisson.pdf](https://github.com/zfit/zfit/blob/develop/tests/test_pdf_poisson.py). Is this intended there because there is more than one function accessing the values? Or is it just legacy and no-one's gotten around to change it?

**Comment by anjabeck** [comment]
Thanks for the comments, I'll change them all.

But for my learning: almost all of the comments refer to stuff that I did because I started from a copy of [test_pdf_poisson.py](https://github.com/zfit/zfit/blob/develop/tests/test_pdf_poisson.py).
Was there a change of conventions (using `znp`, no need for `.numpy()` anymore) and no-one's gotten around to change it? Or is there something peculiar about the poisson vs chi2 that really requires different treatment?

**Comment by jonas-eschle** [comment]
> Thanks for the comments, I'll change them all.

Great, thanks for the contribution!

> But for my learning: almost all of the comments refer to stuff that I did because I started from a copy of [test_pdf_poisson.py](https://github.com/zfit/zfit/blob/develop/tests/test_pdf_poisson.py?rgh-link-date=2024-03-13T10%3A40%3A50Z). Was there a change of conventions (using `znp`, no need for `.numpy()` anymore) and no-one's gotten around to change it? Or is there something peculiar about the poisson vs chi2 that really requires different treatment?

No, you're perfectly right that the Poisson does it "wrong", or less elegant. Over time, things got added and not all tests got the same care of updates

**Comment by jonas-eschle** [comment]
> Please also test in `tests/serialize/test_basics.py`

Good point! You will need to rerun the tests with `--recreate-truth` (once every other test, except the newly created, passes). This will create a new truth file that you can add (that's not yet documented in the contributor guide, updating)

**Comment by anjabeck** [comment]
About the serialization: when adding the chi2 pdf to the set of [basic_pdfs](https://github.com/anjabeck/zfit/blob/chisquared-pdf/tests/serialize/test_basics.py#L318) via something like this
```python
def chisquared(extended=None, **kwargs):
    import zfit

    N = zfit.Parameter("N_chi2", 2)
    obs = zfit.Space("obs", limits=(0, 3))
    return zfit.pdf.ChiSquared(obs=obs,ndof=N,extended=extended)
#     return zfit.pdf.ChiSquared(obs=obs, ndof=N, extended=extended, norm=zfit.Space("norm", limits=(-3, 3)))
```
does not work because there is a mismatch with the other pdfs which are defined in (-3,3).
I tried to add it through defining a new non-zero pdf class:
```python
class chisquared_allownegativex(zfit.pdf.ZPDF):
    _PARAMS = ['ndof']

    def _unnormalized_pdf(self, x):
        x = x.unstack_x(x)
        ndof = self.params['ndof']
        chi2 = zfit.pdf.ChiSquared(obs=self.space, ndof=ndof, extended=self.is_extended)
        return zfit.z.where(x < 0, 0, chi2.pdf(x))


def chisquared(extended=None, **kwargs):
    import zfit

    N = zfit.Parameter("N_chi2", 2)
    obs = zfit.Space("obs", limits=(-3, 3))
    return chisquared_allownegativex(obs=obs, ndof=N, extended=extended)
```
but (probably expectedly) this complains that it doesn't know how to serialize this object
```bash
E           zfit.serialization.serializer.SerializationTypeError: All distributions must be ZfitSerializable
```

I guess one could just change all the range to only allow positive values but maybe that's not ideal.
How are you testing the Poisson right now? I fail to find it here.

**Comment by jonas-eschle** [comment]
The "obs" that has to be the same is an inherent shortcomming of the current serialization, you're absolutely right!

If you would ever want to build a class that uses other PDFs, you should have a look at the Functors, they wrap other pdfs and provide convenient helpers.
This is indeed not quite easy: What should we do if it's < 0, i.e. fail or just set to 0? (I'll check for the poisson tomorrorow what we've done there). What would you expect if you fit something?

For the serialization, I've noted that, so we could also merge without that until the issue is fixed (as serialization is a best-effort basis)

**Comment by jonas-eschle** [comment]
Update, once #523 is merged, this should be fixed (just needs CI to pass)

**Comment by anjabeck** [comment]
> This is indeed not quite easy: What should we do if it's < 0, i.e. fail or just set to 0? (I'll check for the poisson tomorrorow what we've done there). What would you expect if you fit something?

The chi2 and poisson distributions are very well defined mathematical objects with neat properties. These objects are undefined below 0 and I personally think the code should reflect this behaviour and lead to an error.
If a user _really_ needs a distribution that has the shape of a chi2/Poisson but moved to a negative regime, they should build their own pdf. This will not be a chi2/Poisson in the classical sense and may or may not have all the same nice properties but that's the user's problem.

For the serialization, I would expect that is saves the fitting range as well as the normalization range for every individual pdf (both are relevant to rebuild the pdf, no?). I've definitely had real-world (or rather HEP 😅) usecases of fits to a variable in different data sets which required different fit ranges.

**Comment by jonas-eschle** [comment]
> If a user really needs a distribution that has the shape of a chi2/Poisson but moved to a negative regime, they should build their own pdf. This will not be a chi2/Poisson in the classical sense and may or may not have all the same nice properties but that's the user's problem.

Fair, or use a truncated one basically, something like that should work.

> For the serialization, I would expect that is saves the fitting range as well as the normalization range for every individual pdf 

yes, this was simply a bug and fixed now, I've just updated the branch, so make sure to pull first before you push again

**Comment by anjabeck** [comment]
Thanks for merging the bugfix but I'm not sure it solves my original problem: how should I include the chi2 into the serialization tests? I see two options:

1. Test it as part of the basic_pdfs. This would require some gymnastics to create a custom function with normalization range from -3 to 3 (otherwise the sum/product functions complain).
2. Write an entirely new test for pdfs with restricted ranges. This could also include the poisson for which I think there is no serialization test yet.

I think (1) does not make a lot of sense because it wouldn't test the `Repr` object of the chi2 function but rather whatever is used in the custom one. What do you think?

**Comment by jonas-eschle** [comment]
Ah, I see, fair! Actually, you can just make it dynamic, depending on the _original_ pdf space, i.e.
`lower, upper = pdf.space.limit1d` gives you the limits, then sample with those

**Comment by anjabeck** [comment]
> Ah, I see, fair! Actually, you can just make it dynamic, depending on the _original_ pdf space, i.e. `lower, upper = pdf.space.limit1d` gives you the limits, then sample with those

I don't understand how this helps. The current test adds pdfs which are defined on (-3,3). I cannot add `zfit.pdf.ChiSquared` to this test because it cannot have a domain or normalisation range below 0.

**Comment by jonas-eschle** [comment]
> The current test adds pdfs which are defined on (-3,3)

I meant you can make this dynamic that it does _no longer_ use -3, 3, but instead the space of the pdf, which can be different for each pdf. This could potentially have some issues when pdfs autobuild from other pdfs and the obs don't agree, in this case, it's also fine to make it (0, 3) for all, there's nothing inherently peculiar about the negative values.

Otherwise, let me know if this causes too much troubles, I can make the changes

**Comment by anjabeck** [comment]
I reran all tests with `--recreate-truth` and a lot of the binaries changed. Not sure if this is expected?

**Comment by ikrommyd** [comment]
@anjabeck Please run the test without recreating the truth first on your machine. If only the chi-squared related tests fail it's fine.

**Comment by anjabeck** [comment]
> @anjabeck Please run the test without recreating the truth first on your machine. If only the chi-squared related tests fail, it's fine.

> Good point! You will need to rerun the tests with --recreate-truth (once every other test, except the newly created, passes). This will create a new truth file that you can add (that's not yet documented in the contributor guide, updating)

Can you make up your minds about the policy? ;)

**Comment by ikrommyd** [comment]
I think we said the same thing. "once every other test, except the newly created, passes".

**Comment by anjabeck** [comment]
But the other tests cannot pass given the binaries have not been updated to the new range of (-2.1,3) which is nothing I changed but comes from #524

**Comment by ikrommyd** [comment]
Ah I didn't notice the newly created range. I'm sorry then. Binaries generally change whenever I recreate truths. I believe the diff is fine and @jonas-eschle will confirm.

**Comment by anjabeck** [comment]
Ah no worries. I thought so too about the binaries but given that the `json` files had changed as well, I thought it's probably a valid diff?

**Comment by jonas-eschle** [comment]
Thanks for the discussions both of you, the diffs look good, this was indeed my oversight in the other PR, some git stuff slipped through.
Looks good overall, only thing missing seems to be to update the CHANGELOG

**Review by jonas-eschle — APPROVED** [review]
Looks all good!

**Pull request merged**
